File: app/src/main/java/com/sk/revisit/adapter/WebpageItemAdapter.java
package com.sk.revisit.adapter;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.sk.revisit.data.ItemPage;
import com.sk.revisit.databinding.ItemPageBinding;

import java.util.ArrayList;
import java.util.List;

public class WebpageItemAdapter extends RecyclerView.Adapter<WebpageItemAdapter.WebpageItemViewHolder> {

    private List<ItemPage> webPages;
    private List<ItemPage> webPagesOrg;

    public WebpageItemAdapter(List<ItemPage> webPages) {
        this.webPagesOrg = webPages;
        this.webPages = webPages;
    }

    public void setWebPagesOrg(List<ItemPage> webPagesOrg) {
        this.webPagesOrg = webPagesOrg;
        this.webPages = webPagesOrg;
    }

    public void sort() {
        //webPages.sort(ItemPage::getSize);
    }

    public void setWebPages(List<ItemPage> newWebpages) {
        WebpageItemDiffCallback diffCallback = new WebpageItemDiffCallback(webPagesOrg, newWebpages);
        DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(diffCallback);
        webPages = newWebpages;
        diffResult.dispatchUpdatesTo(this);
    }

    public void filter(String query) {
        List<ItemPage> filteredList;
        if (query.trim().isEmpty()) {
            filteredList = webPagesOrg;
        } else {
            filteredList = new ArrayList<>();
            for (ItemPage page : webPagesOrg) {
                if (page.fileName.toLowerCase().contains(query.toLowerCase())) {
                    filteredList.add(page);
                }
            }
        }
        setWebPages(filteredList);
    }

    @NonNull
    @Override
    public WebpageItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemPageBinding binding = ItemPageBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new WebpageItemViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull WebpageItemViewHolder holder, int position) {
        ItemPage page = webPages.get(position);
        holder.bind(page);
    }

    @Override
    public int getItemCount() {
        return webPages.size();
    }

    public static class WebpageItemViewHolder extends RecyclerView.ViewHolder {
        ItemPageBinding binding;

        public WebpageItemViewHolder(ItemPageBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        public void bind(ItemPage page) {
            binding.host.setText(page.host);
            binding.size.setText(page.sizeStr);
            binding.nametext.setText(page.fileName);
        }
    }

    public static class WebpageItemDiffCallback extends DiffUtil.Callback {

        private final List<ItemPage> oldList;
        private final List<ItemPage> newList;

        public WebpageItemDiffCallback(List<ItemPage> oldList, List<ItemPage> newList) {
            this.oldList = oldList;
            this.newList = newList;
        }

        @Override
        public int getOldListSize() {
            return oldList.size();
        }

        @Override
        public int getNewListSize() {
            return newList.size();
        }

        @Override
        public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
            return oldList.get(oldItemPosition).equals(newList.get(newItemPosition));
        }

        @Override
        public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
            return oldList.get(oldItemPosition).equals(newList.get(newItemPosition));
        }
    }
}

File: app/src/main/java/com/sk/revisit/adapter/WebpageItemAdapter.kt
package com.sk.revisit.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.sk.revisit.data.ItemPage
import com.sk.revisit.databinding.ItemPageBinding

class WebpageItemAdapter(
    private var webPagesOrg: List<ItemPage> = emptyList(),
    private val onItemClick: (ItemPage) -> Unit // Added for easy click handling
) : RecyclerView.Adapter<WebpageItemAdapter.ViewHolder>() {

    // The list currently being displayed (filtered)
    private var webPages: List<ItemPage> = webPagesOrg

    fun updateData(newList: List<ItemPage>) {
        val diffCallback = WebpageDiffCallback(webPages, newList)
        val diffResult = DiffUtil.calculateDiff(diffCallback)

        this.webPagesOrg = newList
        this.webPages = newList

        diffResult.dispatchUpdatesTo(this)
    }

    fun filter(query: String) {
        val filteredList = if (query.isBlank()) {
            webPagesOrg
        } else {
            webPagesOrg.filter { it.fileName.contains(query, ignoreCase = true) }
        }

        val diffCallback = WebpageDiffCallback(webPages, filteredList)
        val diffResult = DiffUtil.calculateDiff(diffCallback)

        this.webPages = filteredList
        diffResult.dispatchUpdatesTo(this)
    }

    fun sortBySize() {
        val sortedList = webPages.sortedByDescending { it.size }
        updateDisplayedList(sortedList)
    }

    private fun updateDisplayedList(newList: List<ItemPage>) {
        val diffCallback = WebpageDiffCallback(webPages, newList)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        this.webPages = newList
        diffResult.dispatchUpdatesTo(this)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemPageBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(webPages[position])
    }

    override fun getItemCount(): Int = webPages.size

    inner class ViewHolder(private val binding: ItemPageBinding) : RecyclerView.ViewHolder(binding.root) {

        init {
            binding.root.setOnClickListener {
                val position = bindingAdapterPosition
                if (position != RecyclerView.NO_POSITION) {
                    onItemClick(webPages[position])
                }
            }
        }

        fun bind(page: ItemPage) = with(binding) {
            host.text = page.host
            size.text = page.sizeStr
            nametext.text = page.fileName
        }
    }

    private class WebpageDiffCallback(
        private val oldList: List<ItemPage>,
        private val newList: List<ItemPage>
    ) : DiffUtil.Callback() {

        override fun getOldListSize(): Int = oldList.size
        override fun getNewListSize(): Int = newList.size

        override fun areItemsTheSame(oldPos: Int, newPos: Int): Boolean {
            // Compare unique identifiers (fileName in this case)
            return oldList[oldPos].fileName == newList[newPos].fileName
        }

        override fun areContentsTheSame(oldPos: Int, newPos: Int): Boolean {
            return oldList[oldPos] == newList[newPos]
        }
    }
}